Conway spec prose: GovernanceActions, Voting, Enactment

williamdemeo opened the pull request:
# Description

This addresses three tasks in issue #145.  (GovernanceActions, Voting, and Enactment)

# Checklist

- [X] Commit sequence broadly makes sense and commits have useful messages
- [X] Code is formatted according to [CONTRIBUTING.md](https://github.com/input-output-hk/formal-ledger-specifications/blob/master/CONTRIBUTING.md)
- [X] Self-reviewed the diff

williamdemeo commented:
The items in issue #145 that this PR is meant to address says we "should explain the CC, DReps and give a broad overview of the lifetime of a governance action."  Probably this PR is lacking in at least the "broad overview;" also, the explanations of CC and DReps could probably use clarification and maybe should be expanded upon.

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -112,7 +183,48 @@ NeedsHash (TriggerHF _)        = GovActionID
+the governance action cannot be enacted.
+\begin{itemize}
+\item
+  \defn{Ratification}. An action is said to be \defn{ratified} when it gathers enough votes in its favor (through the rules and parameters that are detailed
```
These can be more specific, we can point to the sections/figures that contain the actual logic.

williamdemeo replied:
I added a reference to the separate section/module on Ratification (\ref{sec:ratification}).

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -112,7 +183,48 @@ NeedsHash (TriggerHF _)        = GovActionID
 NeedsHash Info                 = ⊤
+\end{code}
 
+A governance action is an on-chain event that is triggered by a transaction, and there is a deadline after which
```
Since this document is only about on-chain things we don't need to mention this. Also, an `event` isn't a properly defined thing, so I'd leave that out as well. Here, a governance action is just some data that can be included in a transaction & which can trigger further rules in the ledger.

williamdemeo replied:
Okay, thanks for the clarification.  I'll eliminate "on-chain event."

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -112,7 +183,48 @@ NeedsHash (TriggerHF _)        = GovActionID
+
+\subsection{Voting and ratification}
+\label{sec:voting-and-ratification}
+A governance action is an on-chain event that is triggered by a transaction. There is a deadline after which
```
This sentence and the list below are repeats.

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -112,7 +183,48 @@ NeedsHash (TriggerHF _)        = GovActionID
+\end{itemize}
+
+The data type \AgdaDatatype{Vote} represents the different voting options: \AgdaInductiveConstructor{yes}, \AgdaInductiveConstructor{no}, or \AgdaInductiveConstructor{abstain}.
+Each \AgdaField{vote} is recorded in a \AgdaRecord{GovVote} record along with the following meta-data:
```
Only the `anchor` is metadata, everything else is regular data (since we use it in the rules).

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -131,18 +243,44 @@ record GovProposal : Set where
+\caption{Governance action proposals and votes}
+\label{defs:governance-votes}
+\end{figure*}
+\emph{Any Ada holder} can submit a governance action to the chain.
```
We usually don't mention 'Ada' and 'Lovelace' in the spec. This is also just a marketing line that we don't need to repeat.

williamdemeo replied:
Okay, I removed references to Ada and Lovelace, but kept the explanation about the deposit (required for submitting action to the chain) that will be refunded later... or did you want me to delete that too?

For the same reason I should remove the paragraph,

"As with stake pools, any Ada holder may register to be a DRep and so choose to represent themselves and/or others. Also, as with stake pools, Ada holders may instead delegate their voting rights to any other DRep. Voting rights will be based on the total Ada that is delegated, as a whole number of Lovelace."

that appears near the start of this section, correct?

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -178,25 +319,24 @@ private variable
 instance
   _ = +-0-monoid
-
-data
```
I'd like to keep this part. It looks a lot like the Shelley spec & hides the type theory. We don't want to gatekeep the spec behind an understanding of 'propositions as types'.

WhatisRT reviewed (changes requested):
We should also have a paragraph about `NeedsHash` and `HashProtected`. This is used in ratification for `verifyPrev`, where we check that the stored hash matches the one attached to the action we want to ratify.

We also need to think about how to make the rules for `ENACT` more readable. I find it very hard to parse in its current form.

williamdemeo commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -32,8 +53,53 @@ open import Tactic.Derive.DecEq
+Figure~\ref{defs:governance} defines several data types used to represent governance actions including:
+\begin{itemize}
+  \item %%% \defn{identifier} (\AgdaSymbol{:}\,
+        \AgdaDatatype{GovActionId}---a pair consisting of a transaction id (\AgdaSymbol{:}\,\AgdaDatatype{TxId}) and a natural number (\AgdaSymbol{:}\,\AgdaDatatype{ℕ});
```
I removed some of the type annotations and reworded the explanations, referring to the inhabitant of the type instead of the type.

I also rearranged the file so the code hopefully shows up in the pdf in a more appropriate location. Hopefully this improves the flow and makes it more readable.  

Let me know if you think it still needs improvement and maybe specifically how you would suggest improving it.  Thanks!

williamdemeo commented:
> We should also have a paragraph about `NeedsHash` and `HashProtected`. This is used in ratification for `verifyPrev`, where we check that the stored hash matches the one attached to the action we want to ratify.
> 
> We also need to think about how to make the rules for `ENACT` more readable. I find it very hard to parse in its current form.

I moved `HashProtected` up so it's defined right after `NeedsHash` and added your explanation below the two definitions.  Feel free to suggest a fuller explanation, or I could read up on those types and write a few more words about them.

williamdemeo commented:
The forced push noted above just rebases this PR onto the head of recently updated PR #152.

WhatisRT commented:
I rebased this on `master` so it's easier to review

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -61,7 +90,37 @@ data GovAction : Set where
+\label{defs:governance}
+\end{figure*}
+\begin{table}[h]
+\begin{longtable}[]{@{}
```
I think this table would look a lot better if the constructor name would go on the left (still under the `Action` heading) and the `Agda type` heading was removed. I don't think we need short and long descriptions of every action in a table.

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -61,7 +90,37 @@ data GovAction : Set where
+7. Info & An action that has no effect on-chain, other than an on-chain record. & \AgdaInductiveConstructor{Info}\\
+\hline
+\end{longtable}
+\caption{Types of governance actions}
```
I notice this this is labeled as `Table 2`. Can we just include this in the `figure` counter instead?

williamdemeo replied:
Ummm... not sure why we would do that, but okay.

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -131,23 +221,63 @@ record GovProposal : Set where
+Recall from Section~\ref{sec:protocol-parameters}, the \AgdaRecord{PParam} type is a record containing
+parameters used in the Cardano ledger, grouped according to the general purpose that each parameter serves.
+(See Figure~\ref{fig:protocol-parameter-declarations}.)
+\begin{itemize}
```
We don't need to repeat that list here. The only thing that's relevant in this section is that there are multiple groups, not really their details.

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -131,23 +221,63 @@ record GovProposal : Set where
+\label{sec:protocol-parameters-and-governance-actions}
+Recall from Section~\ref{sec:protocol-parameters}, the \AgdaRecord{PParam} type is a record containing
+parameters used in the Cardano ledger, grouped according to the general purpose that each parameter serves.
+(See Figure~\ref{fig:protocol-parameter-declarations}.)
```
```suggestion
(see Figure~\ref{fig:protocol-parameter-declarations}).
```
And the period above should be removed.

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -131,23 +221,63 @@ record GovProposal : Set where
+which requires an \defn{enact environment} (\AgdaSymbol{:}\,\AgdaRecord{EnactEnv}), an \defn{enact state}
+(\AgdaSymbol{:}\,\AgdaRecord{EnactState}) representing the existing state (prior to enactment), the voted on
+governance action (that achieved enough votes to enact), and the state that results from enacting the given
+governance action (\AgdaSymbol{:}\,\AgdaRecord{EnactState}). (See Figure~\ref{fig:enactment-types}.)
```
```suggestion
governance action (\AgdaSymbol{:}\,\AgdaRecord{EnactState}) (see Figure~\ref{fig:enactment-types}).
```

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -181,22 +314,23 @@ instance
 
-  Enact-NoConf    : ⟦ gid ⟧ᵉ ⊢ s ⇀⦇ NoConfidence            ,ENACT⦈  record s { cc = nothing , gid }
-  Enact-NewComm   : ⟦ gid ⟧ᵉ ⊢ s ⇀⦇ NewCommittee new rem q  ,ENACT⦈ let
+The data type \verb!_⊢_⇀⦇_,ENACT⦈_! defines the transition relation for enacting a governance action.
```
```suggestion
The relation \verb!_⊢_⇀⦇_,ENACT⦈_! is the transition relation for enacting a governance action.
```

WhatisRT reviewed (changes requested):
There are also a lot of type annotations in the text that should be removed

williamdemeo commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -1,5 +1,19 @@
+\end{enumerate}
+Every governance action must be ratified by at least two of these three bodies using their on-chain \defn{votes}.
+The type of action and the state of the governance system determines which bodies must ratify it.
+Ratified actions are then \defn{enacted} on-chain, following a set of well-defined rules.
```
Again, this is probably quoted directly from the CIP... anyway, I'll make the suggested change. 👍🏼

williamdemeo commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -103,7 +162,21 @@ module _ (pp : PParams) (ccThreshold' : Maybe ℚ) where
+\begin{itemize}
+\item
+  \defn{Ratification}. An action is said to be \defn{ratified} when it gathers enough votes in its favor
+  (according to the rules described in Section~\ref{sec:ratification}).
```
Correct, but it will exist once the following PR (#159) gets merged.

williamdemeo commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -113,6 +186,23 @@ NeedsHash (ChangePParams _ _)  = GovActionID
+
+A \defn{governance action proposal} is recorded in a \AgdaRecord{GovProposal} record which includes fields for
+a return address (\FieldTyped{returnAddr}{RwdAddr}), the proposed governance action (\FieldTyped{action}{GovAction}),
+a hash of the previous governance action (\FieldTyped{prevAction}{NeedsHash action}), and an anchor. (See Figure~\ref{defs:governance-votes}.)
```
Although I prefer to write these "see Figure..." hints as separate sentences, it seems you prefer to make them part of the sentence that precedes them, so, for consistency, I'll change all occurrences to use the style you suggest, okay?

WhatisRT replied:
I don't necessarily mind having separate sentences for references (I usually prefer something integrated, but either way is fine), but then they shouldn't have parentheses. If you want parentheses, the main sentence should take care of punctuation. See for example https://www.utsa.edu/twc/documents/Parentheses.pdf.

williamdemeo commented:
I think it looks pretty good now.  There's one unresolved reference which will be resolved as soon as we merge my next PR ( #159 ).

williamdemeo commented:
On Fri, Aug 4, 2023 at 3:43 AM Andre Knispel ***@***.***>
wrote:

> I don't necessarily mind having separate sentences for references (I
> usually prefer something integrated, but either way is fine), but then they
> shouldn't have parentheses. If you want parentheses, the main sentence
> should take care of punctuation. See for example
> https://www.utsa.edu/twc/documents/Parentheses.pdf.
>
From the document you linked to, "It’s also possible to have a
parenthetical that is its own sentence, not encased within a larger
sentence. In this case, it is punctuated like a full and independent
sentence."

williamdemeo commented:
I don't think there are any outstanding change requests, but please correct me if I'm wrong.  Otherwise, LGTM. 👍🏼

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -61,7 +70,48 @@ data GovAction : Set where
+\AgdaInductiveConstructor{NewConstitution}         & a modification to the off-chain Constitution, recorded as an on-chain hash of the text document \\[10pt]
+\AgdaInductiveConstructor{TriggerHF}\footnotemark  & triggers a non-backwards compatible upgrade of the network; requires a prior software upgrade  \\[10pt]
+\AgdaInductiveConstructor{ChangePParams}           & a change to \textit{one or more} updatable protocol parameters, excluding changes to major protocol versions (``hard forks'')\\[10pt]
+\AgdaInductiveConstructor{TreasuryWdrl}            & movements from the treasury, sub-categorized into small, medium or large withdrawals (based on the amount of Lovelace to be withdrawn)\\[10pt]
```
```suggestion
\AgdaInductiveConstructor{TreasuryWdrl}            & movements from the treasury \\[10pt]
```
The CIP is outdated here (and I have an open PR to fix it)

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -103,7 +162,21 @@ module _ (pp : PParams) (ccThreshold' : Maybe ℚ) where
+\end{itemize}
+See Section~\ref{sec:ratification} for more on the ratification process.
 
+\begin{code}
```
The `=` signs still aren't aligned (in the PDF)

WhatisRT commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -131,23 +213,65 @@ record GovProposal : Set where
+\caption{Governance action proposals and votes}
+\label{defs:governance-votes}
+\end{figure*}
+The data type \AgdaDatatype{Vote} represents the different voting options: `yes', `no', or `abstain'.
```
These are constructors too.

WhatisRT reviewed (approved):
I found a few more small things, otherwise LGTM. Please fix those few open things before you merge.

williamdemeo commented:
Oops, I merged this without fixing those small things.  I will make those changes part of the 145-Ratification (#159) PR.  (I've been spending too much time with git merging and rebasing today, so I don't want to do a proper reset in order to make these small changes part of #158 ...sorry.)

williamdemeo commented on src/Ledger/GovernanceActions.lagda:
```diff
@@ -113,6 +171,30 @@ NeedsHash (ChangePParams _ _)  = GovActionID
+  \defn{Ratification}. An action is said to be \defn{ratified} when it gathers enough votes in its favor
+  (according to the rules described in Section~\ref{sec:ratification}).
+\item
+  \defn{Expiration}. An action that doesn't collect sufficient `yes' votes before its deadline is said to have \defn{expired}.
```
Okay, I'll change it back.  I had changed all yes, no and abstains from the Agda colored font to just quoted words, since I got the impression from your earlier comments that we want to make the docs more prosey and less codey... anyway, I'll change them all (or at least some of them) back.

williamdemeo commented:
I've changed `yes`, `no`, and `abstain` from quoted plain-text words to AgdaInductiveConstructors in the GovernanceActions.lagda file.  If you'd like me to do the same in the Ratify.lagda file, please point that out when reviewing PR #159, okay?  Thanks!
